# Исследование надежности заемщиков


## Описание проекта

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку

## Описание данных

- *children* — количество детей в семье
-*days_employed* — общий трудовой стаж в днях
- *dob_years* — возраст клиента в годах
- *education* — уровень образования клиента
- *education_id* — идентификатор уровня образования
- *family_status* — семейное положение
- *family_status_id* — идентификатор семейного положения
- *gender* — пол клиента
- *income_type* — тип занятости
- *debt* — имел ли задолженность по возврату кредитов
- *total_income* — ежемесячный доход
- *purpose* — цель получения кредита

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')


**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

**Рассчитаем долю невозврата кредитов для каждой категории (children)**

In [ ]:
# Построим сводную таблицу, сгруппировав данные по количеству детей
data_pivot = data.pivot_table(index='children', values='debt', aggfunc=['count','sum','mean'])
data_pivot = data_pivot.reset_index()

In [ ]:
# Переименуем столбцы для более читабельного вида
data_pivot.columns = ['children', 'total_loans', 'overdue_loans', 'share_of_overdue_loans']

In [ ]:
data_pivot # Ваш код будет здесь. Вы можете создавать новые ячейки.

children  total_loans  overdue_loans  share_of_overdue_loans
0         0        14091           1063                0.075438
1         1         4808            444                0.092346
2         2         2052            194                0.094542
3         3          330             27                0.081818
4         4           41              4                0.097561
5         5            9              0                0.000000

**Вывод:**
Благодаря сводной таблице отчётливо видно, что
- Многодетные заёмщики не стремятся к долговым обязательствам и реже берут кредиты, чем заёмщики без детей
- Максимальный процент невозврата кредитов у клиентов с 4 детьми - 9.75%
- Количество невозвратов кредитов у клиентов с 1 и 2 ребёнком практически сопоставим
- В среднем процент просроченных должников колеблется от 7% до 10% для каждой категории `children`.

Прямой и однозначной зависимости между количеством детей и возвратом кредита в срок **не наблюдается**. Но если проигнорировать категории людей с 3 и 5 детьми, то можно отметить, что чем больше детей в семье, тем больше трудностей с возвратом кредита. Это довольно логичная тенденциция, т.к увеличивается финансовая нагрузка на семью, что может влиять на способность справляться с долговыми обязательствами.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

**Рассчитаем долю невозврата кредитов для каждой категории (family)**

In [ ]:
family_pivot = data.pivot_table(index='family_status', values='debt', aggfunc=['count','sum','mean'])
family_pivot = family_pivot.reset_index()

In [ ]:
# Переименуем столбцы для более читабельного вида
family_pivot.columns = ['family_status', 'total_loans', 'overdue_loans', 'share_of_overdue_loans']

In [ ]:
family_pivot

family_status  total_loans  overdue_loans  share_of_overdue_loans
0  Не женат / не замужем         2796            273                0.097639
1              в разводе         1189             84                0.070648
2         вдовец / вдова          951             63                0.066246
3       гражданский брак         4134            385                0.093130
4        женат / замужем        12261            927                0.075606

**Вывод:**
Благодаря сводной таблице видно, что
- Максимальный процент невозврата долга у клиентов со статусом "Не женат/ не замужем" - 9.7%;
- Приблизительно высокий показатель(9.3%) у клиентов, живущих в гражданском браке;
- У клиентов со статусом "вдовец / вдова" процент невозврата минимален - 6.6%. Такие клиенты берут кредиты реже остальных (всег 951 человек)

Косвенная зависимость **прослеживается**.
По мере становления человека на его жизненном пути наблюдается рост ответственности.
Наиболее безответственно к погашению кредитов относятся холостые люди и люди в гражданском браке. Более ответственны семейные люди, т.к есть ответственность за благополучие семьи и её устойчивость.
Наиболее ответственна категория "вдовец\вдова", что вполне объяснимо, т.к потеря партнёра вынуждает подходить к финансам критически важно для сохранения стабильности.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
# Построим сводную таблицу, сгруппировав данные по уровню дохода
income_pivot = data.pivot_table(index='total_income_category', values='debt', aggfunc=['count','sum','mean'])
income_pivot = income_pivot.reset_index()

In [ ]:
# Переименуем столбцы для более читабельного вида
income_pivot.columns = ['total_income_category', 'total_loans', 'overdue_loans', 'share_of_overdue_loans']

In [ ]:
income_pivot

total_income_category  total_loans  overdue_loans  share_of_overdue_loans
0                     A           25              2                0.080000
1                     B         5014            354                0.070602
2                     C        15921           1353                0.084982
3                     D          349             21                0.060172
4                     E           22              2                0.090909

**Вывод:**
Благодаря сводной таблице видно, что
- Максимальный процент невозврата долга у клиентов с уровнем дохода категории E (0–30000) - 9.09%;
- Самый низкий % невозврата у клиентов с уровнем дохода категории D (30 001–50 000) - 6.01%.
- У клиентов с уровнем дохода категории A (выше 1 000 001) и C (50 001–200 000) приблизительно одинаковый % не возврата (8 и 8,49 соответственно).

**Вывод**:
Ощутимой зависимости между уровнем дохода и возвратом кредита в срок **явно не прослеживается**, что довольно необычно, т.к более обеспеченные люди не должны иметь проблем с возвратом долгов.
Возможно, в таблице имеют место ошибки в данных при указании дохода клиента. Не исключено, что значения дохода могли указываться в разных валютах, либо за разные промежутки времени, что могло исказить данные о категориях клиентов.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
# Построим сводную таблицу, сгруппировав данные по цели кредита
purpose_pivot = data.pivot_table(index='purpose_category', values='debt', aggfunc=['count','sum','mean'])
purpose_pivot = purpose_pivot.reset_index()

In [ ]:
# Переименуем столбцы для более читабельного вида
purpose_pivot.columns = ['purpose_category', 'total_loans', 'overdue_loans', 'share_of_overdue_loans']

In [ ]:
purpose_pivot

purpose_category  total_loans  overdue_loans  \
0    операции с автомобилем         4279            400   
1  операции с недвижимостью        10751            780   
2     получение образования         3988            369   
3        проведение свадьбы         2313            183   

   share_of_overdue_loans  
0                0.093480  
1                0.072551  
2                0.092528  
3                0.079118

**Вывод:**
Благодаря сводной таблице можно сделать вывод о том, что
- Максимальный процент невозврата кредита с целью операций с автомобилем(9.3%) и для получения образования(9.2%)
- Минимальный процент невозврата кредита для операций с недвижимостью - 7.2%
- Количество клиентов, берущих кредит для операций с недвижимостью больше чем все суммарное количество клиентов, берущих кредит под остальные цели.

**Построим также сводную таблицу по типам категории дохода и целям кредита**

In [ ]:
purpose_income_pivot = data.pivot_table(index=['purpose_category','total_income_category'], values='debt', aggfunc=['count', 'sum', 'mean']).reset_index()
purpose_income_pivot.columns = ['purpose_category','total_income_category', 'total_loans', 'overdue_loans', 'share_of_overdue_loans']
purpose_income_pivot = purpose_income_pivot.sort_values(by=['purpose_category','share_of_overdue_loans'], ascending=[False,False])

In [ ]:
purpose_income_pivot

purpose_category total_income_category  total_loans  \
18        проведение свадьбы                     D           34   
17        проведение свадьбы                     C         1729   
16        проведение свадьбы                     B          544   
15        проведение свадьбы                     A            2   
19        проведение свадьбы                     E            4   
10     получение образования                     A            4   
12     получение образования                     C         3014   
11     получение образования                     B          890   
13     получение образования                     D           75   
14     получение образования                     E            5   
9   операции с недвижимостью                     E           11   
7   операции с недвижимостью                     C         7997   
6   операции с недвижимостью                     B         2558   
5   операции с недвижимостью                     A           17   
8   операции с недвижимостью                     D          168   
3     операции с автомобилем                     D           72   
2     операции с автомобилем                     C         3181   
1     операции с автомобилем                     B         1022   
0     операции с автомобилем                     A            2   
4     операции с автомобилем                     E            2   

    overdue_loans  share_of_overdue_loans  
18              3                0.088235  
17            152                0.087912  
16             28                0.051471  
15              0                0.000000  
19              0                0.000000  
10              1                0.250000  
12            297                0.098540  
11             69                0.077528  
13              2                0.026667  
14              0                0.000000  
9               2                0.181818  
7             599                0.074903  
6             172                0.067240  
5               1                0.058824  
8               6                0.035714  
3              10                0.138889  
2             305                0.095882  
1              85                0.083170  
0               0                0.000000  
4               0                0.000000


**Вывод**:

Автокредиты и кредиты на образование являются одними из самых популярных кредитов по количеству, так и  с точки зрения невозврата банку. Также выяснилось, что наиболее неплатежеспособная категория людей - категория С(50001 - 200000)
Меньше всего задержек на кредиты, связанные с недвижимостью, что вполне объяснимо так как жильё является одной из самой важной жизненной потребностью для человека, требующее осознанного отношения.

В самой малообеспеченной группе Е кредитов на порядок меньше, чем у остальных категорий, что вполне объяснимо, т.к получить кредит при низком доходе затруднительно.


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

**Ответ:**
Вообще, пропуски в данных появляются разными путями. Например, когда клиент не указал о себе какие-нибудь сведения или автоматизированная система сбора информации дала сбой. Иногда пропуски оставляют умышленно, рассчитывая на автозаполнение спецсимволами.
Если ограничиваться данным датасетом и данной задачей, можно предположить, что:
- Пропуски в столбце `total_income` могли взяться из-за сбоя в автоматизированной системе информации, либо из-за нежелания клиента сообщать свой доход;
- Пропуски в столбце `days_employed` могли появиться из-за того, что некоторые заёмщики могли быть безработными

Существует несколько причин отсутствия данных в датасетах, в том числе:<br>
- Человеческий фактор: Ошибки ввода данных, не желание отвечать на определенные вопросы.
- Технические проблемы: Возможны сбои с оборудованием или программным обеспечением для сбора данных. Отсутствующие данные могут быть удалены в процессе обработки или очистки данных.
- Организационные: Конфиденциальность, могут отсутствовать данные, идентифицирующие отдельных лиц или компании.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

**Ответ:**
Пропуски в количественных переменных лучше заполнять именно медианным значением,а не средним арифметическим, потому что среднее арифметическое выборки часто некорректно характеризует данные, например, когда некоторые значения резко выделяются среди большинства. Поэтому медиана - это лучшее решение для количественных переменных при заполнении пропусков.

### Шаг 4: Общий вывод.

В ходе проектной работы сперва было получено базовое представление о данных, осуществлена предобработка данных,а именно:

- обработка пропущенных значений;
- аномальных значений(отрицательные значения столбца `days_employed`;
- обработка неявных дубликатов в столбце `education`;
- удаление дубликатов;
- преобразование типов данных столбцов к удобным для вычисления и анализа;
- категоризация - объединение данных столбцов `total_income` и `purpose` по категориям и создание новых столбцов

Цели получния кредитов были разбиты на укрупненные категории.
Были проанализованы зависимости между данными о заемщике и возвратом кредита в срок

Были получены следующие выводы:
- На погашение кредита в срок не влияет ни семейное положение, ни уровень дохода. Имеет место быть связь с количеством детей и с возрастом заемщиков.
- Чаще всего возвращают в срок кредиты на собственное жилье, а больше всего задержек по кредитам на автомобили и образование.
- Максимальный процент невозврата долга у клиентов со статусом "Не женат/ не замужем" - 9.7% и клиенты, живущие в гражданском браке, а также клиенты с 4 детьми. Данные категории клиентов представляют наибольший риск невозврата средств банку.
- Надежнее выдавать кредиты клиентам с семейным положением "вдова/вдовец", данная категория с большей вероятностью сможет погасить долги. Также вызовет наибольшее доверие клиент, взявший кредит в банке для совершения операций с недвижимостью